### PR TITLE
Manage user roles page

### PR DESCRIPTION
**Pull request by aaronshiel**

### CODE REVIEW

**Review by beatthat — CHANGES_REQUESTED** [review]
looks good. TODO:

- [x] merge main back to this branch (or rebase onto main)
- [x] update gql query names to match `users` not `fetchUsers`
- [x] cut an alpha release docker image
- [x] test locally w latest in 4.2.x-release 

Once the above all confirmed, let me know and I will merge/deply to v2

File: client/src/api.ts [hunk 1]
```diff
@@ -158,6 +159,68 @@ export async function fetchSubjects(
...
+    }`,
+    variables: {
+      filter: JSON.stringify(filter),
+      limit,
+      cursor,
+      sortBy,
+      sortAscending,
+    },
+  });
+  return result.data.data;
```
Comment (beatthat) [inline client/src/api.ts:198]: don't assume GQL requests succeed. You have to check for errors, we should probably have a generic helper function for this

File: client/src/api.ts [hunk 2]
```diff
@@ -158,6 +159,68 @@ export async function fetchSubjects(
...
+          updateUserPermissions(userId: $userId, permissionLevel: $permissionLevel) {
+            userRole
+          }
+        }
+      }`,
+      variables: { userId, permissionLevel },
+    },
+    { headers: headers }
+  );
+  return result.data.data.me.updateUserPermissions;
```
Comment (beatthat) [inline client/src/api.ts:221]: check for errors (have a test for how errors are handled and then make it work here)

**Review by beatthat — CHANGES_REQUESTED** [review]
couple additional notes

File: client/src/api.ts [hunk 3]
```diff
@@ -78,6 +79,28 @@ graphqlRequest.interceptors.response.use(
...
+  }
+  if (res.data.errors) {
+    throw new Error(
+      `errors in GQL response: ${JSON.stringify(res.data.errors)}`
+    );
+  }
+  if (!res.data.data) {
+    throw new Error(`no data in response: ${JSON.stringify(res.data)}`);
+  }
+  if ("me" in res.data.data) return res.data.data.me;
```
Comment (beatthat) [inline client/src/api.ts:100]: the whole `me` thing makes some misery, but we may as well deal w it correctly, e.g.

```typescript
interface Me<T> {
  me: T
}
```

File: client/src/components/nav-bar.tsx [hunk 4]
```diff
@@ -158,7 +161,10 @@ function NavMenu(props: {
   onNav?: (cb: () => void) => void;
 }): JSX.Element {
   const { classes } = props;
-  const { logout } = useWithLogin();
+  const { logout, state } = useWithLogin();
+  const editUsersPermission =
+    state.user?.userRole === UserRole.ADMIN ||
+    state.user?.userRole == UserRole.CONTENT_MANAGER;
```
Comment (beatthat) [inline client/src/components/nav-bar.tsx:167]: `===`

File: client/src/hooks/graphql/use-with-users.tsx [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    reloadData,
+    editData,
+    saveData,
+    sortBy,
+    filter,
+    nextPage,
+    prevPage,
+  } = useWithDataConnection<User>(fetch);
+
+  function fetch() {
```
Comment (beatthat) [inline client/src/hooks/graphql/use-with-users.tsx:37]: needs return type

File: client/src/hooks/graphql/use-with-users.tsx [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    return fetchUsers(searchParams);
+  }
+
+  function onUpdateUserPermissions(userId: string, permissionLevel: string) {
```
Comment (beatthat) [inline client/src/hooks/graphql/use-with-users.tsx:41]: even though i think is just `void` still put the return type

File: client/src/pages/users.tsx [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+  const noEditPermission =
+    props.userRole === UserRole.USER ||
+    (edge.node.userRole === UserRole.ADMIN &&
+      props.userRole !== UserRole.ADMIN);
+
+  function handleRoleChange(user: string, permission: string) {
+    props.userPagin.onUpdateUserPermissions(user, permission);
+  }
+
+  function handleLaunchMentor(mentorId: string) {
```
Comment (beatthat) [inline client/src/pages/users.tsx:143]: ok, just scanning this PR, I've noticed two distinct functions that launch a mentor. Find all such functions and replace them with a single, importable helper function, e.g. `launchMentor(mentorID: string)`

**Review by beatthat — CHANGES_REQUESTED** [review]
looks great. A few very small change requests inline, and then (assuming this is tested locally w an alpha release?) ready to merge/deploy

File: client/src/hooks/graphql/use-with-users.tsx [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+import { Connection, User } from "types";
+import { LoadingError } from "./loading-reducer";
+import {
+  useWithDataConnection,
+  UseDataConnection,
+} from "./use-with-data-connection";
+
+export interface UseUserData extends UseDataConnection<User> {
+  onUpdateUserPermissions: (userId: string, permissionLevel: string) => void;
+  userDataError: LoadingError | undefined;
```
Comment (beatthat) [inline client/src/hooks/graphql/use-with-users.tsx:18]: generally prefer we just use built in typescript for optional types, e.g. `userDataError?: LoadingError`

**Review by beatthat — CHANGES_REQUESTED** [review]
looks great! One or two more tiny change requests inline + sent a PR for `execGQL` update. After those, I think we're good to go

**Comment by aaronshiel** [comment]
tested with alpha release 4.2.0-alpha.21

**Pull request closed without merging**